### PR TITLE
[DO NOT MERGE] Test pynvjitlink 0.2.0 compatibility

### PR DESCRIPTION
**Pull request by isVoid**
Using CI to test compatibility of pynvjitlink 0.2.0. Do not merge.

## Checklist
- [ ] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [ ] New or existing tests cover these changes.
- [ ] The documentation is up to date with these changes.

### CODE REVIEW

File: cpp/cmake/thirdparty/get_arrow.cmake [hunk 1]
```diff
@@ -390,14 +390,11 @@ function(find_and_configure_arrow VERSION BUILD_STATIC ENABLE_S3 ENABLE_ORC ENAB
     BUILD Arrow [=[${CMAKE_CURRENT_LIST_DIR}]=] EXPORT_SET cudf-exports
   )
   rapids_export_find_package_root(
-    BUILD Parquet [=[${CMAKE_CURRENT_LIST_DIR}]=]
-    EXPORT_SET cudf-exports
-    CONDITION ENABLE_PARQUET
+    BUILD Parquet [=[${CMAKE_CURRENT_LIST_DIR}]=] EXPORT_SET cudf-exports CONDITION ENABLE_PARQUET
```
Comment (jakirkham) [inline cpp/cmake/thirdparty/get_arrow.cmake:393]:
```suggestion
    BUILD Parquet [=[${CMAKE_CURRENT_LIST_DIR}]=]
    EXPORT_SET cudf-exports
    CONDITION ENABLE_PARQUET
```

File: cpp/cmake/thirdparty/get_arrow.cmake [hunk 2]
```diff
@@ -390,14 +390,11 @@ function(find_and_configure_arrow VERSION BUILD_STATIC ENABLE_S3 ENABLE_ORC ENAB
...
   )
   rapids_export_find_package_root(
-    BUILD ArrowDataset [=[${CMAKE_CURRENT_LIST_DIR}]=]
-    EXPORT_SET cudf-exports
-    CONDITION ENABLE_PARQUET
+    BUILD ArrowDataset [=[${CMAKE_CURRENT_LIST_DIR}]=] EXPORT_SET cudf-exports CONDITION
+    ENABLE_PARQUET
```
Comment (jakirkham) [inline cpp/cmake/thirdparty/get_arrow.cmake:397]:
```suggestion
    BUILD ArrowDataset [=[${CMAKE_CURRENT_LIST_DIR}]=]
    EXPORT_SET cudf-exports
    CONDITION ENABLE_PARQUET
```

File: cpp/cmake/thirdparty/get_kvikio.cmake [hunk 3]
```diff
@@ -27,9 +27,7 @@ function(find_and_configure_kvikio VERSION)
 
   include("${rapids-cmake-dir}/export/find_package_root.cmake")
   rapids_export_find_package_root(
-    BUILD KvikIO "${KvikIO_BINARY_DIR}"
-    EXPORT_SET cudf-exports
-    CONDITION KvikIO_BINARY_DIR
+    BUILD KvikIO "${KvikIO_BINARY_DIR}" EXPORT_SET cudf-exports CONDITION KvikIO_BINARY_DIR
```
Comment (jakirkham) [inline cpp/cmake/thirdparty/get_kvikio.cmake:30]:
```suggestion
    BUILD KvikIO "${KvikIO_BINARY_DIR}"
    EXPORT_SET cudf-exports
    CONDITION KvikIO_BINARY_DIR
```

File: cpp/cmake/thirdparty/get_kvikio.cmake [hunk 4]
```diff
@@ -1,5 +1,5 @@
 # =============================================================================
-# Copyright (c) 2022-2023, NVIDIA CORPORATION.
+# Copyright (c) 2022-2024, NVIDIA CORPORATION.
```
Comment (jakirkham) [inline cpp/cmake/thirdparty/get_kvikio.cmake:2]:
```suggestion
# Copyright (c) 2022-2023, NVIDIA CORPORATION.
```

**Review by jakirkham — COMMENTED** [review]
Thanks Michael! 🙏

Looks like CI likes the wrapped lines we had before. Included some suggestions below to go back to that

There were also some CI fixes that have since been merged. Made sure those are merged back into this PR

Let's see how these changes go

File: dependencies.yaml [hunk 5]
```diff
@@ -559,7 +559,8 @@ dependencies:
           - matrix: {cuda: "12.*"}
             packages:
               - rmm-cu12==24.6.*
-              - pynvjitlink-cu12
+              - --extra-index-url=https://pypi.anaconda.org/rapidsai-wheels-nightly/simple
```
Comment (jakirkham) [inline dependencies.yaml:562]: Think we already add this. It is where the RAPIDS 24.06 packages are coming from as well

Let's try dropping and see if things still work. We can always revisit if not

```suggestion
```

**Comment by jakirkham** [comment]
Looks like we are having trouble picking up `libkvikio` packages in [this CI job]( https://github.com/rapidsai/cudf/actions/runs/8699225939/job/23857956394?pr=15540#step:7:337 ):

```
Could not solve for environment specs
The following package could not be installed
└─ libcudf-tests is not installable because it requires
   └─ libcudf 24.06.00a113 cuda12_240416_g768a984370_113, which requires
      └─ libkvikio 24.06.* , which does not exist (perhaps a missing channel).
```

However [the `libkvikio` version `24.06` nightly packages do exist]( https://anaconda.org/rapidsai-nightly/libkvikio/files?version=24.06.00a )

**Comment by isVoid** [comment]
> Looks like we are having trouble picking up `libkvikio` packages in [this CI job](https://github.com/rapidsai/cudf/actions/runs/8699225939/job/23857956394?pr=15540#step:7:337):
> 
> ```
> Could not solve for environment specs
> The following package could not be installed
> └─ libcudf-tests is not installable because it requires
>    └─ libcudf 24.06.00a113 cuda12_240416_g768a984370_113, which requires
>       └─ libkvikio 24.06.* , which does not exist (perhaps a missing channel).
> ```
> 
> However [the `libkvikio` version `24.06` nightly packages do exist](https://anaconda.org/rapidsai-nightly/libkvikio/files?version=24.06.00a)

The x86 test looks to be passing:
https://github.com/rapidsai/cudf/actions/runs/8699225939/job/23857956270?pr=15540

**Comment by jakirkham** [comment]
Looks like this may still be picking up the old `pynvjitlink`

**Comment by isVoid** [comment]
> Looks like this may still be picking up the old `pynvjitlink`

Ah, the `pynvjitlink` lib is not marked as test dependency, but only as runtime dependency. In the test environment the pynvjitlink is pulled via a transitive dependency. I wonder if this is intended? @brandon-b-miller

**Comment by brandon-b-miller** [comment]
I believe we should have it in the testing environment too, yes. There's [tests in cuDF](https://github.com/rapidsai/cudf/blob/branch-24.06/python/cudf/cudf/tests/test_mvc.py) that assume it's presence.

**Comment by isVoid** [comment]
I tried to add a `specific` dependency item for pynvjitlink but it seems like I made a mistake:
```
Traceback (most recent call last):
  File "/opt/conda/bin/rapids-dependency-file-generator", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.10/site-packages/rapids_dependency_file_generator/cli.py", line 171, in main
    make_dependency_files(parsed_config, args.config, to_stdout)
  File "/opt/conda/lib/python3.10/site-packages/rapids_dependency_file_generator/rapids_dependency_file_generator.py", line 451, in make_dependency_files
    raise ValueError(
ValueError: No matching matrix found in 'test_python_cudf' for: {'cuda': '12.0', 'arch': 'aarch64', 'py': '3.10'}
```

**Comment by jakirkham** [comment]
Seeing an interesting error in [the CUDA 12 ARM Python 3.10 test job]( https://github.com/rapidsai/cudf/actions/runs/8707235994/job/23883829952?pr=15540#step:7:27 ):

```python
Traceback (most recent call last):
  File "/opt/conda/bin/rapids-dependency-file-generator", line 10, in <module>
    sys.exit(main())
  File "/opt/conda/lib/python3.10/site-packages/rapids_dependency_file_generator/cli.py", line 171, in main
    make_dependency_files(parsed_config, args.config, to_stdout)
  File "/opt/conda/lib/python3.10/site-packages/rapids_dependency_file_generator/rapids_dependency_file_generator.py", line 451, in make_dependency_files
    raise ValueError(
ValueError: No matching matrix found in 'test_python_cudf' for: {'cuda': '12.0', 'arch': 'aarch64', 'py': '3.10'}
```

Do we know why this is coming up?

**Comment by bdice** [comment]
Commit 89bcda2c079a4a728d6f3a4abf4857c9d2a5de10 should fix the issues here.

Explanations:

There was an empty matrix like

```yaml
- matrix:
  packages:
```

that was deleted in `dependencies.yaml`. We needed that empty matrix to provide a match for ARM (though it provides no packages).

There were also a bunch of unnecessary dependency entries in the test environment. These should already be provided as run dependencies of the pip/conda packages, so we shouldn't need to list them explicitly.

Finally, the conda recipe needed its pinning updated to `pynvjitlink=0.2.0`. This should test what you were hoping to test.

**Comment by bdice** [comment]
> Ah, the `pynvjitlink` lib is not marked as test dependency, but only as runtime dependency. In the test environment the pynvjitlink is pulled via a transitive dependency. I wonder if this is intended? @brandon-b-miller

This is expected behavior -- pynvjitlink _is_ a runtime dependency, and all runtime dependencies will be installed when you install cudf into the test environment. Test dependencies are only for packages that are not required by cudf but are needed to run its tests (e.g. `pytest`).

**Comment by vyasr** [comment]
I'm going to close this since we've been using newer versions of pynvjitlink safely for a while now.

**Pull request closed without merging**